port_forwarding.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		return err
+	}
+	sshdC.port = mappedPort.Port()
+
+	sshConfig := ssh.ClientConfig{
+		User:            user,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Auth:            []ssh.AuthMethod{ssh.Password(sshPassword)},
+		Timeout:         30 * time.Second,
+	}
+
+	sshdC.sshConfig = &sshConfig
+
+	return nil
+}
+
+func (sshdC *sshdContainer) exposeHostPort(port int) error {
+	if _, ok := sshdC.tunnels[port]; ok {
+		// do not expose the same port twice
+		return nil
+	}
+
+	// Setup the tunnel, but do not yet start it yet.
+	tunnel := newSSHTunnel(
+		fmt.Sprintf("%s@localhost:%s", user, sshdC.port),
+		sshdC.sshConfig,
+		"localhost", port, // The destination host and port of the actual server.
+	)
+
+	// use testcontainers logger
+	tunnel.Log = Logger
+
+	errorGr := errgroup.Group{}
+	errorGr.Go(func() error {
+		return tunnel.Start()
+	})
+
+	if err := errorGr.Wait(); err != nil {
+		return fmt.Errorf("failed to start the SSH tunnel: %w", err)
+	}
+
+	sshdC.tunnels[port] = tunnel
+
+	return nil
+}
+
+type sshEndpoint struct {
+	Host string
+	Port int
+	User string
+}
+
+func newSshEndpoint(s string) *sshEndpoint {
+	endpoint := &sshEndpoint{
+		Host: s,
+	}
+	if parts := strings.Split(endpoint.Host, "@"); len(parts) > 1 {
+		endpoint.User = parts[0]
+		endpoint.Host = parts[1]
+	}
+	if parts := strings.Split(endpoint.Host, ":"); len(parts) > 1 {
+		endpoint.Host = parts[0]
+		endpoint.Port, _ = strconv.Atoi(parts[1])
+	}
+	return endpoint
+}
+
+func (endpoint *sshEndpoint) String() string {
+	return fmt.Sprintf("%s:%d", endpoint.Host, endpoint.Port)
+}
+
+type ssHTunnel struct {
+	Local  *sshEndpoint
+	Server *sshEndpoint
+	Remote *sshEndpoint
+	Config *ssh.ClientConfig
+	Log    Logging
+}
+
+func newSSHTunnel(tunnel string, sshConfig *ssh.ClientConfig, target string, targetPort int) *ssHTunnel {
+	destination := fmt.Sprintf("%s:%d", target, targetPort)
+
+	localEndpoint := newSshEndpoint(destination)
+
+	server := newSshEndpoint(tunnel)
+	if server.Port == 0 {
+		server.Port = 22
+	}
+
+	return &ssHTunnel{
+		Config: sshConfig,
+		Local:  localEndpoint,
+		Server: server,
+		Remote: newSshEndpoint(destination),
+	}
+}
+
+func (tunnel *ssHTunnel) logf(fmt string, args ...interface{}) {
+	if tunnel.Log != nil {
+		tunnel.Log.Printf(fmt, args...)
+	}
+}
+
+func (tunnel *ssHTunnel) Close() error {
+	if tunnel.Log != nil {
+		tunnel.logf("closing tunnel")
+	}
+
+	return nil
+}
+
+func (tunnel *ssHTunnel) Start() error {
+	listener, err := net.Listen("tcp", tunnel.Local.String())
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+
+	tunnel.Local.Port = listener.Addr().(*net.TCPAddr).Port
+	for {
+		conn, err := listener.Accept()
+		if err != nil {
+			return err
+		}
+		tunnel.logf("accepted connection")
+		go tunnel.forward(conn)
+	}
+}
+
+func (tunnel *ssHTunnel) forward(localConn net.Conn) {
+	serverConn, err := ssh.Dial("tcp", tunnel.Server.String(), tunnel.Config)
```

Comment:
maybe DialContext so that an incoming context with a deadline is paid attention to? Same below

```suggestion
	serverConn, err := ssh.DialContext("tcp", tunnel.Server.String(), tunnel.Config)
```

Reply:
I do not see Dial context in the latest release of the crypto package: https://cs.opensource.google/go/x/crypto/+/master:ssh/client.go;l=176 Is it not merged yet?

Reply:
my mistake.. I must have looked at the wrong type at the time